ssm/main.tf

```diff
@@ -1,64 +1,277 @@
+  bucket = local.s3_bucket_name
+  acl    = "private"
+
+  logging {
+    target_bucket = local.log_bucket
+    target_prefix = "${local.s3_bucket_name}/"
+  }
+
+  tags = {
+    environment = var.env_name
+  }
+
+  server_side_encryption_configuration {
+    rule {
+      apply_server_side_encryption_by_default {
+        kms_master_key_id = aws_kms_key.kms_ssm.key_id
+        sse_algorithm     = "aws:kms"
+      }
+    }
+  }
+
+  versioning {
+    enabled = true
+  }
+
+  lifecycle_rule {
+    id      = "expire"
+    prefix  = "/"
+    enabled = true
+
+    transition {
+      storage_class = "INTELLIGENT_TIERING"
+    }
+    noncurrent_version_transition {
+      storage_class = "INTELLIGENT_TIERING"
+    }
+    expiration {
+      days = 2190
+    }
+    noncurrent_version_expiration {
+      days = 2190
+    }
+  }
+}
+
+module "ssm_logs_bucket_config" {
+  source = "github.com/18F/identity-terraform//s3_config?ref=7e11ebe24e3a9cbc34d1413cf4d20b3d71390d5b"
+
+  bucket_name_prefix   = var.bucket_name_prefix
+  bucket_name          = "${var.env_name}-ssm-logs"
+  region               = var.region
+  inventory_bucket_arn = "arn:aws:s3:::${local.inventory_bucket}"
+}
+
+resource "aws_cloudwatch_log_group" "ssm_session_logs" {
+  name              = "aws-ssm-sessions-${var.env_name}" #stream name must start with "aws-ssm-logs"
+  retention_in_days = 365
+  kms_key_id        = aws_kms_key.kms_ssm.arn
+}
+
+# Policy and roles to permit SSM access / actions on EC2 instances, and to allow them to send metrics and logs to CloudWatch
+data "aws_iam_policy_document" "ssm_access_role_policy" {
+  # Basic
   statement {
-    sid = "ssmmessages"
+    sid = "SSMCoreAccess"
     actions = [
+      "ssm:UpdateInstanceInformation",
       "ssmmessages:CreateControlChannel",
       "ssmmessages:CreateDataChannel",
       "ssmmessages:OpenControlChannel",
-      "ssmmessages:OpenDataChannel",
-      "ssm:UpdateInstanceInformation",
+      "ssmmessages:OpenDataChannel"
     ]
-
     resources = [
       "*",
     ]
   }
   statement {
-    sid = "ssmlogs"
+    sid = "CloudWatchLogsAccessForSSM"
     actions = [
-      "logs:CreateLogStream",
-      "logs:PutLogEvents",
+      "logs:CreateLogGroup",
       "logs:DescribeLogGroups",
       "logs:DescribeLogStreams",
+      "logs:PutLogEvents",
+    ]
+    resources = [
+      "*"
+    ]
+  }
+  # S3
+  statement {
+    sid = "S3ConfigAccessForSSM"
+    actions = [
+      "s3:GetEncryptionConfiguration"
     ]
-
     resources = [
       "*",
     ]
   }
+  statement {
+    sid = "S3LoggingAccessForSSM"
+    actions = [
+      "s3:PutObject"
+    ]
+    resources = [
+      "${aws_s3_bucket.ssm_logs.arn}/*"
+    ]
+  }
+  # KMS
+  statement {
+    sid = "KMSDecryptionAccess"
+    actions = [
+      "kms:Decrypt",
+    ]
+    resources = [
+      aws_kms_key.kms_ssm.arn
+    ]
+  }
+
+  statement {
+    sid = "KMSDataKeyAccess"
+    actions = [
+      "kms:GenerateDataKey",
+    ]
+    resources = [
```

Comment:
Can you reduce the scope to just the key in question?   (Docs seem to indicate it can accept a specific symmetric KMS instance.)

Reply:
Moved and updated!